### PR TITLE
Edit captain/front back

### PR DESCRIPTION
**Pull request by naremanhilles**
### Related:
related to #58 front end
#69 back end
#81 query 

### Summary:
A couple of sentences explaining the contents of the PR

### Notes:
Optional notes

### Checklist Criteria
- [ ] **Branch is up to date with master**
- [ ] **The project runs**
- - [ ] You can download and install locally, and run the project locally
- - [ ] All existing and new features run without bugs
- - [ ] The build script runs without error
- [ ] **Code is consistent**
- - [ ] Linter is used, no error messages
- - [ ] Code is consistent with agreed team style
- - [ ] Naming of files is consistent with project (e.g. all files named with camelCase)
- [ ] **No console logs**
- - [ ] Console.log is only used for crucial system messages (e.g. which port server is on).
- - [ ] Any console.logs used for debugging have been removed.

### CODE REVIEW

File: client/src/Components/Layouts/Captains/EditCaptain/index.jsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  render() {
+    return (
+      <div>
+        <CollectionsPage />
+      </div>
+    );
+  }
+}
+
+export default EditCaptain;
```
Comment (denaHS) [inline client/src/Components/Layouts/Captains/EditCaptain/index.jsx]: add a new line!!!

File: client/src/Components/Layouts/Captains/EditCaptain/style.css [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ }
+ .edit-captain__popup .anticon svg{
+   display: none;
+ }
+.edit-captain__popup .ant-modal-content {
+  width: 130%;
+ right: calc(50% - 166px);
+}
+.edit-captain__popup .ant-modal-header{
+  /* padding: 16px 0px;
```
Comment (denaHS) [inline client/src/Components/Layouts/Captains/EditCaptain/style.css]: delete all commint

File: client/src/Components/Layouts/Captains/EditCaptain/style.css [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ .image__name{
+  position: absolute;
+  width: fit-content;
+} 
+.upload-photo{
+  border: 1px solid #ddd;
+  border-radius: 5px;
+  padding: 7px;
+  width: 115px;
+}
```
Comment (denaHS) [inline client/src/Components/Layouts/Captains/EditCaptain/style.css]: add new line

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+          if (res.result) {
+            const rows = res.result[0];
+            const status = rows.b_status ? "true" : "false";
+            this.props.form.setFieldsValue({
+              name: rows.s_name, email: rows.s_email, phone: rows.s_mobile_number,
+              address: rows.s_address, IDNumber: rows.s_id_number, licenceNumber: rows.s_driver_licence_number,
+              status, file: rows.s_image,
+            });
+            var formData = this.props.form.getFieldsValue();
+            fetch(`/api/v1/image/${rows.s_image}`)
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles It might be easier to read if you moved this image fetch (line 37-48)  into its own function and called that function here.

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+              .then(res => res.arrayBuffer())
+              .then(response => {
+                let typeArray = new Uint8Array(response);
+                const stringChar = String.fromCharCode.apply(null, typeArray);
+
+                this.setState({
+                  photoUrl: stringChar
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles is stringChar a dataUrl?
I wonder if it would work if you directly set the photoUrl to `/api/v1/image/${rows.s_image}`?

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+      const uploadButton = (
+        <Button className="btn--upload">
+
+          <Icon type={this.state.loading ? 'check-circle' : 'upload'} />
+          أرفق صورة
+        </Button>
+
+      );
+      const imageUrl = this.state.imageUrl;
+      const photoUrl = this.state.photoUrl;
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles If you want, you can shorten this to `const { imageUrl, photoUrl } = this.state`.  It's one less line :) which might help since this is a long file.

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+                  accept=".jpg , .png, .jpeg"
+                  name="file"
+                  className="avatar-uploader"
+                  showUploadList={false}
+                  customRequest={this.handleReuest}
+                  multiple={false}
+                >
+                  {uploadButton}
+                  <div className="image__name">
+                    {(imageUrl) ? imageUrl && <><Icon type="check-circle" />{imageUrl.name}</> : photoUrl && <><Icon /><img src={photoUrl} className="upload-photo" /></>}
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles To make this more readable, you could create a few helper functions, and then change the line to call them:
`{ imageUrl ? this.renderImageUrl() : this.renderPhotoUrl() }`
and then somewhere else in your class you can add the helper functions:
`renderImageUrl = () => <><Icon type="check-circle" />{this.state.imageUrl.name}</>`
`renderPhotoUrl = () => this.state.photoUrl && <><Icon /></>`

Also you can see that I removed the `imageUrl &&` because the first part of that line (the `imageUrl ?` is already ensuring that imageUrl is truthy.

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                  ]
+                })(<Input type="password" id="password" />)}
+              </Form.Item>
+              <Form.Item label="الهاتف">
+                {getFieldDecorator("phone", {
+                  rules: [
+                    {
+                      required: true,
+                      message: "الرجاء ملئ الحقل بارقام",
+                      pattern: /^\+?[0-9]{10,12}$/
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles it might make your users' life easier if you allow dashes, spaces, and dots in the phone numbers too, so they can enter things like this:
+1 123-456-7890
+1 1234 56 789
+33 123.456.7890

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+                    }
+                  ]
+                })(<Input type="text" id="phone" />)}
+              </Form.Item>
+
+              <Form.Item label="العنوان" dir="ltr">
+                {getFieldDecorator("address", {
+                  rules: [
+                    { required: true, message: "يرجى ملئ الحقل بحروف ", pattern: /^[أ-يa-z]*$/ }
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles Also here it might make users less frustrated if the regex is relaxed to let people add different special characters.  Address fields are hard :)
Someone might want to enter an address like this:
123 N. 5th St. #5
New York NY 10009

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+  handleCancel = () => {
+    this.setState({ visible: false });
+  };
+
+
+  handleCreate = () => {
+    const form = this.formRef.props.form;
+    form.validateFields((err, values) => {
+      if (err) {
+        return;
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles You could use notification.open() to display the error here (maybe you're already planning to do this in the future)

File: client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+        formData.append('address', address);
+        formData.append('email', email);
+        formData.append('licenceNumber', licenceNumber);
+        formData.append('name', name);
+        formData.append('password', password);
+        formData.append('phone', phone);
+        formData.append('status', status);
+        const id = 6;
+        // const { id } = this.props.id;
+        fetch(`/api/v1/putCaptain/${id}`, {
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/editCaptain.jsx]: @naremanhilles It's not something you need to change right now, but an idea for next time: It might be better code structure to have one component responsible for all of the network requests.
Right now, CollectionsPage does the PUT to the backend, but CollectionCreateForm does the GET to the backend.  Ideally, one of them would be doing both PUT and GET, and the other wouldn't have to worry about doing network requests.

File: client/src/Components/Layouts/Captains/EditCaptain/index.jsx [hunk 12]
```diff
@@ -0,0 +1,16 @@
+
+import React, { Component } from 'react';
+import CollectionsPage from './editCaptain';
+
+// eslint-disable-next-line react/prefer-stateless-function
+class EditCaptain extends Component {
```
Comment (tomduggan85) [inline client/src/Components/Layouts/Captains/EditCaptain/index.jsx]: @naremanhilles I have to agree with eslint here 😅.  It's not super urgent, but it would be a little more efficient to have this small component be a functional component instead of a class component.  Then the browser doesn't have to use memory to keep track of the class instance.

File: server/database/queries/captain/editCaptain.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+const connection = require('../../config/dbConnection');
+
+const editCaptain = (captainInfo, id) => {
+  const queryValues = Object.keys(captainInfo).map(info => captainInfo[info]);
```
Comment (tomduggan85) [inline server/database/queries/captain/editCaptain.js]: @naremanhilles I think this could give you an unexpected order of keys in the future, and break the query.  It might be better to be explicit:
`const { name, email, mobile_number, .............. } = captainInfo`

Then you could be clearer in your if-statement about which value you are checking:
`if ( !password ) {`

File: client/src/Components/Layouts/Captains/EditCaptain/index.jsx (hunk 12 above)
Comment (tomduggan85) [reply client/src/Components/Layouts/Captains/EditCaptain/index.jsx]: haha, thanks ✨eslint✨
